segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -136,52 +137,188 @@ public void initialize(long epoch) {
             conf.set("dfs.client.block.write.replace-datanode-on-failure.policy", "NEVER");
         }
 
-        this.context = new FileSystemOperation.OperationContext(epoch, openFileSystem(conf), this.config);
+        this.context = new OperationContext(epoch, openFileSystem(conf), this.config);
         log.info("Initialized (HDFSHost = '{}', Epoch = {}).", this.config.getHdfsHostURL(), epoch);
     }
 
-    protected FileSystem openFileSystem(Configuration conf) throws IOException {
-        return FileSystem.get(conf);
+    @Override
+    public SegmentProperties getStreamSegmentInfo(String streamSegmentName) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "getStreamSegmentInfo", streamSegmentName);
+        SegmentProperties result = null;
+        int attemptCount = 0;
+        do {
+            long length = 0;
+            boolean isSealed = false;
+            FileStatus last = null;
+            try {
+                last = findStatusForSegment(streamSegmentName, true);
+                isSealed = isSealed(last.getPath());
+            } catch (IOException fnf) {
+                // This can happen if we get a concurrent call to SealOperation with an empty last file; the last file will
+                // be deleted in that case so we need to try our luck again (in which case we need to refresh the file list).
+                if (++attemptCount < MAX_ATTEMPT_COUNT) {
+                    continue;
+                }
+                HDFSExceptionHelpers.throwException(streamSegmentName, fnf);
+            }
+
+            result = StreamSegmentInformation.builder().name(streamSegmentName).length(last.getLen()).sealed(isSealed).build();
+        } while (result == null);
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName, result);
+        return result;
     }
 
     @Override
-    public SegmentProperties create(String streamSegmentName) throws StreamSegmentException {
-        return call(new CreateOperation(streamSegmentName, this.context));
+    public boolean exists(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "exists", streamSegmentName);
+        FileStatus status;
+        try {
+            status = findStatusForSegment(streamSegmentName, false);
+        } catch (IOException e) {
+            return false;
```

Comment:
This may be just a personal preference but I prefer to have a single exit. So you could do like:
```
+        FileStatus status = null;
+        try {
+            status = findStatusForSegment(streamSegmentName, false);
+        } catch (IOException e) {
+            log.warn("Got exception checking if file exists",e);
+        }
+        boolean exists = status != null;
+        LoggerHelpers.traceLeave(log, "exists", traceId, streamSegmentName, exists);
+        return exists;